edu_edfi_airflow/dags/edfi_resource_dag.py

```diff
@@ -277,195 +356,458 @@ def build_change_version_task_group(self) -> TaskGroup:
+        # This should NOT be necessary, but we encountered a bug where a downstream "none_skipped" task skipped with "upstream_failed" status.
+        def fail_if_xcom(xcom_value, **context):
+            if xcom_value:
+                raise AirflowFailException
+            else:
+                raise AirflowSkipException  # Force a skip to not mark the taskgroup as a success when all tasks skip.
+
+        failed_sentinel = PythonOperator(
+            task_id=f"{task_id}__failed_total_counts",
+            python_callable=fail_if_xcom,
+            op_args=[airflow_util.xcom_pull_template(get_cv_operator, key='failed_endpoints')],
             trigger_rule='all_done',
             dag=self.dag
         )
+        get_cv_operator >> failed_sentinel
 
+        return get_cv_operator
 
-    def build_dbt_var_increment_operator(self):
+    def build_change_version_update_operator(self,
+        task_id: str,
+        endpoints: List[str],
+        get_deletes: bool,
+        get_key_changes: bool,
+        **kwargs
+    ) -> PythonOperator:
         """
 
         :return:
         """
-        def short_circuit_update_variable(**kwargs):
-            """
-            Helper to build short-circuit logic into the update_variable callable if no new data was ingested.
-            :return:
-            """
-            from airflow.exceptions import AirflowSkipException
-
-            for task_id in kwargs['task'].upstream_task_ids:
-                if kwargs['ti'].xcom_pull(task_id):
-                    return update_variable(**kwargs)
-            else:
-                raise AirflowSkipException(
-                    "There is no new data to process using DBT. All upstream tasks skipped or failed."
-                )
-
         return PythonOperator(
-            task_id='increment_dbt_variable',
-            python_callable=short_circuit_update_variable,
+            task_id=task_id, 
+            python_callable=change_version.update_change_versions,
             op_kwargs={
-                'var': self.dbt_incrementer_var,
-                'value': lambda x: int(x) + 1,
+                'tenant_code': self.tenant_code,
+                'api_year': self.api_year,
+                'snowflake_conn_id': self.snowflake_conn_id,
+                'change_version_table': self.change_version_table,
+                
+                'edfi_change_version': airflow_util.xcom_pull_template(self.newest_edfi_cv_task_id),
+                'endpoints': endpoints,
+                'get_deletes': get_deletes,
+                'get_key_changes': get_key_changes,
             },
-            trigger_rule='all_done',
-            dag=self.dag
+            provide_context=True,
+            dag=self.dag,
+            **kwargs
         )
 
 
-    def build_edfi_to_snowflake_task_group(self,
-        resource : str,
-        namespace: str = 'ed-fi',
+    # Polymorphic Ed-Fi TaskGroups
+    @staticmethod
+    def xcom_pull_template_map_idx(task_ids, idx: int):
```

Comment:
These XCom helpers are used throughout the task groups to retrieve elements from the list of tuples. They use the new XCom Pull Template methods.